core/trino-main/src/main/java/io/trino/util/StatementUtils.java

```diff
@@ -233,11 +238,60 @@ private StatementUtils() {}
             .build().stream()
             .collect(toImmutableMap(StatementTypeInfo::getStatementType, identity()));
 
-    public static Optional<QueryType> getQueryType(Statement statement)
+    private static final String INFORMATION_SCHEMA = "information_schema";
+    private static final String SYSTEM_JDBC = "system.jdbc";
+    private static final String SYSTEM_METADATA = "system.metadata";
+    private static final String SYSTEM_RUNTIME = "system.runtime";
+
+    private static final ImmutableSet<String> METADATA_CATALOG_SCHEMAS = ImmutableSet.<String>builder()
+            .add(SYSTEM_JDBC)
+            .add(SYSTEM_METADATA)
+            .add(SYSTEM_RUNTIME)
+            .build();
+
+    public static Optional<QueryType> getQueryType(Session session, Statement statement)
     {
         if (statement instanceof ExplainAnalyze) {
-            return getQueryType(((ExplainAnalyze) statement).getStatement());
+            return getQueryType(session, ((ExplainAnalyze) statement).getStatement());
         }
+
+        if (statement instanceof Query) {
+            boolean isMetadataQuery = false;
+
+            HashSet<Table> tables = getTables(statement, new HashSet<String>());
+
+            if (tables.isEmpty()) {
+                isMetadataQuery = true;
+            }
+            else {
+                isMetadataQuery = true;
+                for (Table t : tables) {
+                    String[] tableParts = t.getName().toString().split("\\.");
+                    String catalog = "";
+                    String schema = "";
+                    if (tableParts.length == 3) {
+                        catalog = tableParts[0];
+                        schema = tableParts[1];
+                    }
+                    else if (tableParts.length == 2 && session.getCatalog().isPresent()) {
+                        catalog = session.getCatalog().get();
+                        schema = tableParts[0];
+                    }
+                    else if (tableParts.length == 1 && session.getCatalog().isPresent() && session.getSchema().isPresent()) {
+                        catalog = session.getCatalog().get();
+                        schema = session.getSchema().get();
+                    }
```

Comment:
This would be cleaner IMO if the logic to qualify the table name were extracted to a separate method. E.g. `QualifiedName qualifyTableName(Session session, Table table)`

Reply:
Made the suggested change.